server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java

```diff
@@ -0,0 +1,138 @@
+ * due to the master node not understanding cluster state with transport versions added in 8.8.0.
+ * Any nodes with the inferred placeholder cluster state is then refreshed with their actual transport version
+ */
+public class TransportVersionsFixupListener implements ClusterStateListener {
+
+    private static final Logger Log = LogManager.getLogger(TransportVersionsFixupListener.class);
+
+    private static final TransportVersion INFERRED_VERSION = TransportVersion.V_8_8_0;
+
+    private final MasterServiceTaskQueue<NodeTransportVersionTask> taskQueue;
+    private final ClusterAdminClient client;
+    private final Set<String> pendingNodes = Collections.synchronizedSet(new HashSet<>());
+
+    public TransportVersionsFixupListener(ClusterService service, ClusterAdminClient client) {
+        this(service.createTaskQueue("fixup-transport-versions", Priority.LOW, new TransportVersionUpdater()), client);
+    }
+
+    TransportVersionsFixupListener(MasterServiceTaskQueue<NodeTransportVersionTask> taskQueue, ClusterAdminClient client) {
+        this.taskQueue = taskQueue;
+        this.client = client;
+    }
+
+    record NodeTransportVersionTask(Map<String, TransportVersion> results) implements ClusterStateTaskListener {
+        @Override
+        public void onFailure(Exception e) {
+            Log.error("Could not apply transport version for nodes {} to cluster state", results.keySet(), e);
+        }
+    }
+
+    private static class TransportVersionUpdater implements ClusterStateTaskExecutor<NodeTransportVersionTask> {
+        @Override
+        public ClusterState execute(BatchExecutionContext<NodeTransportVersionTask> context) throws Exception {
+            ClusterState.Builder builder = ClusterState.builder(context.initialState());
+            for (var c : context.taskContexts()) {
+                for (var e : c.getTask().results().entrySet()) {
+                    // this node's transport version might have been updated already/node has gone away
+                    if (Objects.equals(builder.transportVersions().get(e.getKey()), INFERRED_VERSION)) {
+                        builder.putTransportVersion(e.getKey(), e.getValue());
+                    }
+                }
+            }
+            return builder.build();
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster() == false) return; // only if we're master
+
+        // if the min node version > 8.8.0, and the cluster state has some transport versions == 8.8.0,
+        // then refresh all inferred transport versions to their real versions
+        // now that everything should understand cluster state with transport versions
+        if (event.state().nodes().getMinNodeVersion().after(Version.V_8_8_0)
+            && event.state().getMinTransportVersion().equals(INFERRED_VERSION)) {
```

Comment:
I think we should be explicit about this not being a coincidence:

```suggestion
        if (event.state().nodes().getMinNodeVersion().after(INFERRED_VERSION)
            && event.state().getMinTransportVersion().equals(INFERRED_VERSION)) {
```

Could we also use the same constant in `org.elasticsearch.cluster.ClusterState#inferTransportVersion`?

Reply:
That's different types, the INFERRED_VERSION is a TransportVersion

Reply:
Oh I see, sorry yes wrong types.

But yes +1 to a public static constant.